src/Compilers/Core/Portable/Compilation/Compilation.cs

```diff
@@ -1488,6 +1489,161 @@ void addIfNotNull(INamedTypeSymbol? toAdd)
             ImmutableArray<bool> memberIsReadOnly,
             ImmutableArray<NullableAnnotation> memberNullableAnnotations);
 
+        /// <summary>
+        /// Creates an <see cref="IMethodSymbol"/> whose <see cref="IMethodSymbol.MethodKind"/> is <see
+        /// cref="MethodKind.BuiltinOperator"/> for a binary operator. Built-in operators are commonly created for
+        /// symbols like <c>bool int.operator==(int v1, int v2)</c> which the language implicitly supports, even if such
+        /// a symbol is not explicitly defined for that type in either source or metadata.
+        /// </summary>
+        /// <param name="name">The binary operator name.  Should be one of the names from <see cref="WellKnownMemberNames"/>.</param>
+        /// <param name="returnType">The return type of the binary operator.</param>
+        /// <param name="leftType">The type of the left operand of the binary operator.</param>
+        /// <param name="rightType">The type of the right operand of the binary operator.</param>
+        /// <param name="isChecked">Whether or not this is a checked binary operator.  For example <c>int int.operator
+        /// checked +(int v1, int v2)</c>.</param>
+        /// <returns></returns>
+        public IMethodSymbol CreateBuiltinOperator(string name, ITypeSymbol returnType, ITypeSymbol leftType, ITypeSymbol rightType, bool isChecked)
+        {
+            // Can't check 'name' here as VB and C# support a different subset of names.
+
+            if (returnType is null)
+                throw new ArgumentNullException(nameof(returnType));
+
+            if (leftType is null)
+                throw new ArgumentNullException(nameof(leftType));
+
+            if (rightType is null)
+                throw new ArgumentNullException(nameof(rightType));
+
+            // When the types are dynamic, then virtually all operators are allowed and synthesized.
+            // So only check when neither are dynamic.
+            if (leftType.TypeKind != TypeKind.Dynamic && rightType.TypeKind != TypeKind.Dynamic)
+            {
+                CheckSupportedBinaryOperatorType(returnType, nameof(returnType));
+                CheckSupportedBinaryOperatorType(leftType, nameof(leftType));
+                CheckSupportedBinaryOperatorType(rightType, nameof(rightType));
+            }
+
+            return CommonCreateBuiltinOperator(name, returnType, leftType, rightType, isChecked);
```

Comment:
>CommonCreateBuiltinOperator [](http://example.com/codeflow?start=19&length=27)

How hard would it be to filter operator names based on type involved? For example, logical and is not supported for integers, etc. Also some operators for special types do not allow type mix. For example, there is no operator ```int+byte```. Though rules could be language specific

Reply:
> How hard would it be to filter operator names based on type involved? 

My main concerns there would be:

1. having to duplicate all the logic already in the compiler elsewhere in some of yoru tables (though if you can help by giving a suggestion on how to potentially just use those tables to validate things, that would make this more palatable).
2. potentially having this logic start to diverge from the actual compiler logic (Again, if we can reuse the actual logic... this gets palatable).

A good example of this might be the utf8 string work.  Imagine if this helper were written prior to that.  Do we think we would have effectively fixed this up when that was added, or we could have missed it?  I'd prefer to be a little more lenient than restrictive.  However, i don't feel strongly and would be willing to beef thigns up here (though i could use some help with that).

Reply:
Note: this is the allure of the "allow everything" approach.  Sure, it allows for completely bogus methods to be created.  But that doesn't particularly bother me.  Then tehre is no maintenance parity concern for future changes.  I can see the desire to not allow wonky things though, but i'm wary about how complex it will be to support that.  If there's some nice way to do it though, i'm def supportive.

Reply:
For C#, compiler tables are more overload resolution tables. I.e., given types, give me an operator, if any. Not so much a list of supported operators. However, pretty much for all operators both parameter types should be the same (shift operators probably the only exception), and throwing this type to the overload resolution table should yield back the same type. Here are the methods that do table driven overload resolution:  [https://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Binder/Semantics/Operators/BinaryOperatorEasyOut.cs,8bb194cd6f262730](https://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Binder/Semantics/Operators/BinaryOperatorEasyOut.cs,8bb194cd6f262730), [https://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Binder/Semantics/Operators/UnaryOperatorEasyOut.cs,1dd7a6290245c3c0](https://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Binder/Semantics/Operators/UnaryOperatorEasyOut.cs,1dd7a6290245c3c0). It looks like it should be possible to use them as is, or create clones that operate with different input types.

Reply:
It looks like VB goes the same way, perhaps with additional exceptions. Functions to look at for overload resolution: [https://sourceroslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Semantics/Operators.vb,333d336f7b1c2a60](https://sourceroslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Semantics/Operators.vb,333d336f7b1c2a60), [https://sourceroslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Semantics/Operators.vb,f6c0997d84c5b913](https://sourceroslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Semantics/Operators.vb,f6c0997d84c5b913).